File: ruby/cal_points.rb
# frozen_string_literal: true

# @param {String[]} operations
# @return {Integer}
def cal_points(operations)
  stack = []

  operations.each do |operation|
    case operation
    when 'C'
      stack.pop
    when 'D'
      stack.push(stack[-1] * 2)
    when '+'
      stack.push(stack[-1] + stack[-2])
    else
      stack.push(operation.to_i)
    end
  end

  stack.sum
end

File: typescript/calPoints.ts
function calPoints(operations: string[]): number {
  const scores: number[] = [];

  for (const operation of operations) {
    if (operation === "+") {
      const score = scores[scores.length - 1] + scores[scores.length - 2];
      scores.push(score);
    } else if (operation === "D") {
      const score = 2 * scores[scores.length - 1];
      scores.push(score);
    } else if (operation === "C") {
      scores.pop();
    } else {
      const score = parseInt(operation, 10);
      scores.push(score);
    }
  }

  return scores.reduce((sum, score) => sum + score, 0);
}
